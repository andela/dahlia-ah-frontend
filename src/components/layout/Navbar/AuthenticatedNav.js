import React, { useEffect } from 'react';
import Materialize from 'materialize-css';

const AthenticatedNav = () => {
  useEffect(() => {
    const dropdown = document.querySelector('.dropdown-trigger');
    const notification = document.querySelector('.notification');
    Materialize.Dropdown.init(dropdown);
    Materialize.Dropdown.init(notification);
  }, []);

  return (
    <div className="navbar-fixed">
      <nav className="authenticated-nav navbar">
        <div className="nav-wrapper">
          <img
            className="brand-logo center"
            src="https://res.cloudinary.com/drlcfqzym/image/upload/v1567007007/ah-logo-black.png"
            alt="Author's Haven"
          />
          <a href="#!" type="button" data-target="mobile-demo" className="sidenav-trigger">
            <img src="https://img.icons8.com/ios/50/000000/menu.png" alt="menu" className="font-icon" />
          </a>
          <button type="button" className="nav-btn btn-small hide-on-med-and-down">upgrade</button>
          <ul id="nav-mobile" className="right hide-on-med-and-down">
            <li className="notification" data-target="notification-dropdown">
              <img
                src="https://res.cloudinary.com/drlcfqzym/image/upload/v1567080014/notification_icon.png"
                alt="notification icon"
              />
            </li>
            <ul id="notification-dropdown" className="dropdown-content">
              <li>Welcome</li>
            </ul>
            <li className="dropdown-trigger" data-target="nav-dropdown">
              <img
                src="https://res.cloudinary.com/drlcfqzym/image/upload/v1567080039/person_icon.png"
                alt="profile"
              />
              <span className="greeting">
                Hi, James
              </span>
              <img
                src="https://img.icons8.com/ios/20/000000/expand-arrow--v1.png"
                alt="dropdown"
              />
            </li>
            <ul id="nav-dropdown" className="dropdown-content">
              <li className="nav-link">Account</li>
              <li className="nav-link">Settings</li>
              <li className="nav-link">Help</li>
              <li className="nav-link">Sign Out</li>
            </ul>
          </ul>
        </div>
      </nav>
      <ul className="sidenav" id="mobile-demo">
        <li>Account</li>
        <li>Settings</li>
        <li>Help</li>
        <li>Sign Out</li>
      </ul>
    </div>
  );
};

export default AthenticatedNav;
