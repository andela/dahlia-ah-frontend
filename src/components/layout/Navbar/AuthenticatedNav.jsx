/* eslint-disable no-restricted-globals */
import { Link, withRouter } from 'react-router-dom';
import React, { useEffect, useState, useContext } from 'react';
import PropTypes from 'prop-types';
import useAuthenticatedNav from './effects/AuthenticatedNav';
import { UserContext } from '../../../contexts/UserContext';
import axios from '../../../api';
import SignOut from '../../helpers/SignOut';
import socket from '../../../helpers/socket';
import notificationState from '../../../hooks/Notification';
import Notification from '../../Notification/Notification';
import { markAllAsRead } from '../../../api/notification';
import { AuthModalContext } from '../../../context/AuthModalContext';

const AuthenticatedNav = ({ history }) => {
  const { id } = JSON.parse(window.localStorage.getItem('AuthorsHavenUser'));
  const { notifications, setNotification } = notificationState();
  const [isOpen, setIsOpen] = useState(false);
  const { canWrite } = useContext(AuthModalContext);


  const handleInstantNotification = (data) => {
    setNotification([data, ...notifications]);
  };

  const getUnread = () => {
    const unread = notifications.filter((notification) => notification.isRead === false);
    return unread.length;
  };

  socket.on(`notification-${id}`, (data) => {
    handleInstantNotification(data);
  });

  // get notification everytime auth nav mount
  useEffect(() => {
    document.body.addEventListener('click', (event) => {
      const { target: { id: elementId } } = event;
      const dontAffect = ['notItem', 'not-collection', 'notification', 'mark-all', 'notText', 'notImage'];
      if (!dontAffect.find((blacklistId) => blacklistId === elementId)) {
        setIsOpen(false);
      }
    });
    axios.get('/notifications')
      .then(({ data: { notifications: userNotifications } }) => {
        setNotification(userNotifications);
      })
      .catch(() => {});
  }, []);

  useAuthenticatedNav();

  return (
    <UserContext.Consumer>
      {(context) => {
        const { firstName, isVerified } = context.user;

        const startWritingBtn = canWrite ? (
          <button type="button" className="nav-btn btn-small hide-on-med-and-down" onClick={() => { history.push('/write-novel'); }}>
            Start Writing
          </button>
        ) : (
          <button type="button" className="nav-btn btn-small hide-on-med-and-down">
            upgrade
          </button>
        );

        return (
          <div className="navSection">
            <div className="navbar-fixed">
              <nav className={isVerified || location.pathname !== '/confirmation-page' ? 'authenticated-nav navbar' : 'authenticated-nav navbar confirm-page'}>
                <div className="nav-wrapper">
                  <img
                    className="brand-logo center"
                    src="https://res.cloudinary.com/drlcfqzym/image/upload/v1567007007/ah-logo-black.png"
                    alt="Author's Haven"
                  />
                  <a href="#!" type="button" data-target="mobile-demo" className="sidenav-trigger">
                    <img src="https://img.icons8.com/ios/50/000000/menu.png" alt="menu" className="font-icon" />
                  </a>
                  {startWritingBtn}
                  <ul id="nav-mobile" className="right hide-on-med-and-down">
                    <li className="notification" onClick={() => setIsOpen(!isOpen)} onKeyDown={() => setIsOpen(false)} role="presentation">
                      <img
                        src="https://img.icons8.com/ios/25/000000/appointment-reminders.png"
                        alt="notification icon"
                        id="notification"
                      />
                      { notifications.length > 0 && getUnread() > 0 ? <div className="notification-badge">{ getUnread() < 100 ? getUnread() : '99+' }</div> : null }
                    </li>
                    {
                      isOpen ? (
                        <ul className="notification-content">
                          <Notification
                            markAllAsRead={markAllAsRead}
                            notifications={notifications}
                            setNotification={setNotification}
                          />
                        </ul>
                      ) : null
                    }

                    <li className="dropdown-trigger" data-target="nav-dropdown">
                      <img
                        src="https://img.icons8.com/color/25/000000/person-male.png"
                        alt="profile"
                      />
                      <span className="greeting">
                        Hi,
                        { ` ${firstName}` }
                      </span>
                      <img
                        src="https://img.icons8.com/ios/20/000000/expand-arrow--v1.png"
                        alt="dropdown"
                      />
                    </li>
                    <ul id="nav-dropdown" className="dropdown-content">
                      <li className="nav-link">
                        <Link to={{
                          pathname: '/profile',
                        }}
                        >
                        Profile
                        </Link>

                      </li>
                      <li className="nav-link"><a href="#!">My Novels</a></li>
                      <li className="nav-link"><a href="#!">Bookmarks</a></li>
                      <li className="nav-link">
                        <Link to={{
                          pathname: '/user/settings',
                        }}
                        >
                          Settings
                        </Link>
                      </li>
                      <li className="nav-link"><a href="#!">Help</a></li>
                      <li className="nav-link"><a href="/" onClick={SignOut}>Sign Out</a></li>
                    </ul>
                  </ul>
                </div>
              </nav>
              <ul className="sidenav" id="mobile-demo">
                { canWrite ? (
                  <li className="nav-link">
                    <Link to={{
                      pathname: '/write-novel',
                    }}
                    >
                    Start Writing
                    </Link>
                  </li>
                ) : null }
                <li className="nav-link">
                  <Link to={{
                    pathname: '/profile',
                  }}
                  >
                        Profile
                  </Link>

                </li>
                <li className="nav-link"><a href="#!">My Novels</a></li>
                <li className="nav-link"><a href="#!">Bookmarks</a></li>
                <li className="nav-link">
                  <Link to={{
                    pathname: '/user/settings',
                  }}
                  >
                  Settings
                  </Link>
                </li>
                <li className="nav-link"><a href="#!">Help</a></li>
                <li className="nav-link"><a href="/" onClick={SignOut}>Sign Out</a></li>
              </ul>
            </div>
          </div>
        );
      }}

    </UserContext.Consumer>
  );
};

AuthenticatedNav.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func,
  }).isRequired,
};

export default withRouter(AuthenticatedNav);
