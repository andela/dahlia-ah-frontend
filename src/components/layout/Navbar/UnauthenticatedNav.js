import React from 'react';
import { Link } from 'react-router-dom';

const UnauthenticatedNav = () => (
  <div className="navbar-fixed">
    <nav className="unauthenticated-nav navbar">
      <div className="nav-wrapper">
        <img
          className="brand-logo center"
          src="https://res.cloudinary.com/drlcfqzym/image/upload/v1566998517/ah-logo-white.png"
          alt="Author's Haven"
        />
        <a href="#!" type="button" data-target="mobile-demo" className="sidenav-trigger">
          <img src="https://img.icons8.com/ios/50/000000/menu.png" alt="menu" className="font-icon" />
          {/* <i className="fas fa-bars" />/ */}
        </a>
        <ul id="nav-mobile" className="right hide-on-med-and-down">
          <Link to="/login"><li className="nav-link">Sign Up</li></Link>
          <Link to="/login"><li className="nav-btn">Sign In</li></Link>
        </ul>
      </div>
    </nav>
    <ul className="sidenav" id="mobile-demo">
      <Link to="/login"><li className="nav-link">Sign Up</li></Link>
      <Link to="/login"><li className="nav-link">Sign In</li></Link>
    </ul>
  </div>
);

export default UnauthenticatedNav;
