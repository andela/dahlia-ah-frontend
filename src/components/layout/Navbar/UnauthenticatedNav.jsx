import React, { useContext } from 'react';
import { AuthModalContext } from '../../../context/AuthModalContext';
import Login from '../../Login/Login';
import SignUpFormContainer from '../../SignupForm/SignupFormContainer';

const UnauthenticatedNav = () => {
  const { modalComponent, setModalComponent } = useContext(AuthModalContext);

  const handleOpenModal = (component) => {
    setModalComponent(component);
    document.body.style.height = '100vh';
    document.body.style.overflowY = 'hidden';
  };

  const handleCloseModal = () => {
    setModalComponent(null);
    document.body.style.height = '100vh';
    document.body.style.overflowY = 'auto';
  };

  let renderedModal = null;
  switch (modalComponent) {
    case 'signup':
      renderedModal = (
        <SignUpFormContainer
          closeModal={handleCloseModal}
          openModal={handleOpenModal}
        />
      );
      break;
    case 'signin':
      renderedModal = (
        <Login
          closeModal={handleCloseModal}
          openModal={handleOpenModal}
        />
      );
      break;
    default:
      break;
  }
  return (
    <>
      { renderedModal }
      <div className="navSection">
        <div className="navbar-fixed">
          <nav className="unauthenticated-nav navbar">
            <div className="nav-wrapper">
              <img
                className="brand-logo center"
                src="https://res.cloudinary.com/drlcfqzym/image/upload/v1566998517/ah-logo-white.png"
                alt="Author's Haven"
              />
              <a href="#!" type="button" data-target="mobile-demo" className="sidenav-trigger">
                <img src="https://img.icons8.com/ios/50/ffffff/menu.png" alt="menu" className="font-icon" />
              </a>
              <ul id="nav-mobile" className="right hide-on-med-and-down">
                <button type="button" className="nav-link" onClick={() => { handleOpenModal('signup'); }}>Sign Up </button>
                <button type="button" className="nav-btn" onClick={() => { handleOpenModal('signin'); }}>Sign In</button>
              </ul>
            </div>
          </nav>
          <ul className="sidenav" id="mobile-demo">
            <li>
              <button type="button" className="nav-link" onClick={() => { handleOpenModal('signup'); }}>Sign Up </button>
            </li>
            <li>
              <button type="button" className="nav-btn" onClick={() => { handleOpenModal('signin'); }}>Sign In</button>
            </li>
          </ul>
        </div>
      </div>
    </>
  );
};

export default UnauthenticatedNav;
