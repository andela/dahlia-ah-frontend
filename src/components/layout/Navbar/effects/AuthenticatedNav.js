import { useEffect } from 'react';
import Materialize from 'materialize-css';

const useAuthenticatedNav = () => useEffect(() => {
  const dropdown = document.querySelector('.dropdown-trigger');
  Materialize.Dropdown.init(dropdown);
}, []);

export default useAuthenticatedNav;
