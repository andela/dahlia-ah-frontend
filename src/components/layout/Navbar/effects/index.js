import { useEffect } from 'react';
import Materialize from 'materialize-css';

const useNavigation = () => useEffect(() => {
  const authNav = document.querySelector('.authenticated-nav');
  const nav = document.querySelector('.navbar');

  const sidenav = document.querySelector('.sidenav');
  Materialize.Sidenav.init(sidenav);

  let changeNavColor = 'nav-black';
  if (authNav) {
    changeNavColor = 'nav-white';
  }
  document.addEventListener('scroll', () => {
    const scroll = window.scrollY;
    if (scroll > 10) {
      nav.classList.add('change-nav', `${changeNavColor}`);
    } else {
      nav.classList.remove('change-nav', `${changeNavColor}`);
    }
  });
}, []);

export default useNavigation;
