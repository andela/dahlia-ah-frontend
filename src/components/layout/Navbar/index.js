import React, { useEffect } from 'react';
import Materialize from 'materialize-css';
import './navbar.scss';
import UnauthenticatedNav from './UnauthenticatedNav';

const Navbar = () => {
  useEffect(() => {
    const sidenav = document.querySelector('.sidenav');
    Materialize.Sidenav.init(sidenav);

    const nav = document.querySelector('.navbar');
    const authNav = document.querySelector('.authenticated-nav');
    let changeNavColor = 'nav-black';
    if (authNav) {
      changeNavColor = 'nav-white';
    }
    document.addEventListener('scroll', () => {
      const scroll = window.scrollY;
      if (scroll > 100) {
        nav.classList.add('change-nav', `${changeNavColor}`);
      } else {
        nav.classList.remove('change-nav', `${changeNavColor}`);
      }
    });
  }, []);

  return (
    <>
      <UnauthenticatedNav />
    </>
  );
};

export default Navbar;
