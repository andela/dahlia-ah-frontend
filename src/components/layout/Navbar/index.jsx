import React, { useEffect } from 'react';
import { withRouter } from 'react-router-dom';
import ReactRouterPropTypes from 'react-router-prop-types';
import useNavigation from './effects';
import './navbar.scss';
import UserContextProvider from '../../../contexts/UserContext';
import AuthenticatedNav from './AuthenticatedNav';
import UnauthenticatedNav from './UnauthenticatedNav';

const Navbar = ({ history }) => {
  useNavigation();
  let user = JSON.parse(window.localStorage.getItem('AuthorsHavenUser'));
  const path = history.location.pathname;
  useEffect(() => {
    user = JSON.parse(window.localStorage.getItem('AuthorsHavenUser'));
  }, [history]);

  return (
    <>
      {
        user && user.token && path !== '/'
          ? (
            <UserContextProvider>
              <AuthenticatedNav />
            </UserContextProvider>
          )
          : <UnauthenticatedNav />
      }
    </>
  );
};

Navbar.propTypes = {
  history: ReactRouterPropTypes.history.isRequired,
};

export default withRouter(Navbar);
