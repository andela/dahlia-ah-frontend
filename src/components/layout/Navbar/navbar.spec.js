import React from 'react';
import { BrowserRouter, Switch } from 'react-router-dom';
import { configure, mount, shallow } from 'enzyme';

import Adapter from 'enzyme-adapter-react-16';
import Navbar from './index';
import AuthModalContextProvider from '../../../context/AuthModalContext';
import AuthenticatedNav from './AuthenticatedNav';
import UnauthenticatedNav from './UnauthenticatedNav';

configure({ adapter: new Adapter() });

describe('App component', () => {
  it('renders correctly', () => {
    const wrapper = mount(
      <BrowserRouter>
        <Switch>
          <AuthModalContextProvider>
            <Navbar />
          </AuthModalContextProvider>
        </Switch>
      </BrowserRouter>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    const wrapper = shallow(<AuthenticatedNav />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    const wrapper = shallow(
      <AuthModalContextProvider>
        <UnauthenticatedNav />
      </AuthModalContextProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
