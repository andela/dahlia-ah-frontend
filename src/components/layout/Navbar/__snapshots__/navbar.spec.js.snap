// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders correctly 1`] = `ReactWrapper {}`;

exports[`App component renders without crashing 1`] = `ShallowWrapper {}`;

exports[`App component renders without crashing 2`] = `ShallowWrapper {}`;
